README
# Flask-Restless #

## Introduction ##

This is Flask-Restless, a [Flask][1] extension that creates URL endpoints that
satisfy the requirements of the [JSON API][2] specification. It is compatible
with models that have been defined using either [SQLAlchemy][3] or
[Flask-SQLAlchemy][4].

This document contains some brief instructions concerning installation of
requirements, installation of this extension, configuration and usage of this
extension, and building of documentation.

For more information, see the

  * [documentation][5],
  * [Python Package Index listing][6],
  * [source code repository][7].

[![Build status](https://secure.travis-ci.org/jfinkels/flask-restless.png)](https://travis-ci.org/jfinkels/flask-restless)

[1]: http://flask.pocoo.org
[2]: https://jsonapi.org
[3]: https://sqlalchemy.org
[4]: https://packages.python.org/Flask-SQLAlchemy
[5]: https://flask-restless.readthedocs.org
[6]: https://pypi.python.org/pypi/Flask-Restless
[7]: https://github.com/jfinkels/flask-restless

## Copyright license ##

The code comprising this program is copyright 2011 Lincoln de Sousa and
copyright 2012, 2013, 2014, 2015, 2016 Jeffrey Finkelstein and contributors,
and is dual-licensed under the following two copyright licenses:

* the GNU Affero General Public License, either version 3 or (at your option)
  any later version
* the 3-clause BSD License

For more information, see the files `LICENSE.AGPL` and `LICENSE.BSD` in this
directory.

The documentation is licensed under the Creative Commons Attribution-ShareAlike
4.0 license.

## Contents ##

This is a partial listing of the contents of this package.

* `LICENSE.AGPL` - one possible copyright license under which this program is
  distributed to you (the GNU Affero General Public License version 3)
* `LICENSE.BSD` - another possible copyright license under which this program
  is distributed to you (the 3-clause BSD License)
* `docs/` - the Sphinx documentation for Flask-Restless
* `examples/` - example applications of Flask-Restless
* `flask_restless/` - the Python package containing the extension
* `MANIFEST.in` - indicates files to include when packaging Flask-Restless
* `README.md` - this file
* `setup.py` - Python setuptools configuration file for packaging this
  extension
* `tests/` - unit tests for Flask-Restless

The `flask_restless` directory is a Python package containing the following
files and directory:

* `helpers.py` - utility functions, mainly for performing introspection on
  SQLAlchemy objects
* `manager.py` - contains the main class that end users will utilize to create
  ReSTful JSON APIs for their database models
* `search.py` - functions and classes that facilitate searching the database
  on requests that require a search
* `serialization.py` - basic serialization and deserialization for SQLAlchemy
  models
* `views/` - the view classes that implement the JSON API interface

## Installing ##

This application can be used with any Python version that Flask supports, which
currently includes versions 2.6, 2.7, 3.3, and 3.4. (Python 3.2 is not
supported by Flask and therefore cannot be supported by Flask-Restless.)

This application requires the following libraries to be installed:

* [Flask][1] version 0.10 or greater
* [SQLAlchemy][3] version 0.8 or greater
* [python-dateutil][8] version strictly greater than 2.2
* [mimerender][9] version 0.5.2 or greater

These requirements (and some additional optional packages) are also listed in
the `requirements/install.txt` file. Using `pip` is probably the easiest way to
install these:

    pip install -r requirements/install.txt

[8]: https://labix.org/python-dateutil
[9]: https://github.com/martinblech/mimerender

## Building as a Python egg ##

This package can be built, installed, etc. as a Python egg using the provided
`setup.py` script. For more information, run

    python setup.py --help

## How to use ##

For information on how to use this extension, build the documentation here or
[view it on the Web][5].

## Testing ##

If your Python interpreter is `cpython`, run:

    pip install -r requirements/test-cpython.txt

Otherwise, if your Python interpreter is `pypy`, run:

    pip install -r requirements/test-pypy.txt

To run the tests:

    python setup.py test

## Building documentation ##

Flask-Restless requires the following program and supporting library to build
the documentation:

* [Sphinx][10]
* [sphinxcontrib-httpdomain][11], version 1.1.7 or greater

These requirements are also listed in the `requirements/doc.txt` file. Using
`pip` is probably the easiest way to install these:

    pip install -r requirements/doc.txt

The documentation is written for Sphinx in [reStructuredText][12] files in the
`docs/` directory. Documentation for each class and function is provided in the
docstring in the code.

The documentation uses the Flask Sphinx theme. It is included as a git
submodule of this project, rooted at `docs/_themes`. To get the themes, do

    git submodule update --init

Now to build the documentation, run the command

    python setup.py build_sphinx

in the top-level directory. The output can be viewed in a web browser by
opening `build/sphinx/html/index.html`.

[10]: http://sphinx.pocoo.org/
[11]: https://packages.python.org/sphinxcontrib-httpdomain/
[12]: https://docutils.sourceforge.net/rst.html

## Contributing ##

Please report any issues on the [GitHub Issue Tracker][13].

To suggest a change to the code or documentation, please create a new pull
request on GitHub. Contributed code must come with an appropriate unit
test. Please ensure that your code follows [PEP8][14], by running, for example,
[flake8][15] before submitting a pull request. Also, please squash multiple
commits into a single commit in your pull request by [rebasing][16] onto the
master branch.

By contributing to this project, you are agreeing to license your code
contributions under both the GNU Affero General Public License, either version
3 or any later version, and the 3-clause BSD License, and your documentation
contributions under the Creative Commons Attribution-ShareAlike License version
4.0, as described in the copyright license section above.

[13]: http://github.com/jfinkels/flask-restless/issues
[14]: https://www.python.org/dev/peps/pep-0008/
[15]: http://flake8.readthedocs.org/en/latest/
[16]: https://help.github.com/articles/about-git-rebase/

## Artwork ##

The `artwork/flask-restless-small.svg` and
`docs/_static/flask-restless-small.png` are licensed under the [Creative
Commons Attribute-ShareAlike 4.0 license][17]. The original image is a scan of
a (now public domain) illustration by Arthur Hopkins in a serial edition of
"The Return of the Native" by Thomas Hardy published in October 1878.

The `artwork/flask-restless.svg` and `docs/_static/flask-restless.png` are
licensed under the [Flask Artwork License][18].

[17]: https://creativecommons.org/licenses/by-sa/4.0
[18]: http://flask.pocoo.org/docs/license/#flask-artwork-license

## Contact ##

Jeffrey Finkelstein <jeffrey.finkelstein@gmail.com>


CHANGES
Changelog
=========

Here you can see the full list of changes between each Flask-Restless release.
Version 1.0.0 saw a major overhaul of Flask-Restless to make it compliant with
JSON API, so changes from prior versions may not be relevant to more recent
versions.

Numbers following a pound sign (#) refer to `GitHub issues`_.

.. _GitHub issues: https://github.com/jfinkels/flask-restless/issues

Version 1.0.0b2-dev
-------------------

*This is a beta release; these changes will appear in the 1.0.0 release.*

Not yet released.

- Eliminates all documentation build warnings for bad references.
- Changes serialization/deserialization to class-based implementation instead
  of a function-based implementation. This also adds support for serialization
  of heterogeneous collections.
- Removes `mimerender`_ as a dependency.
- :issue:`7`: allows filtering before function evaluation.
- :issue:`49`: deserializers now expect a complete JSON API document.
- :issue:`200`: be smarter about determining the ``collection_name`` for
  polymorphic models defined with single-table inheritance.
- :issue:`253`: don't assign to callable attributes of models.
- :issue:`268`: adds top-level endpoint that exposes API schema.
- :issue:`479`: removes duplicate (and sometimes conflicting)
  :http:header:`Content-Type` header in responses.
- :issue:`481,488`: added negation (``not``) operator for search.
- :issue:`492`: support JSON API recommended "simple" filtering.
- :issue:`508`: flush the session before postprocessors, and commit after.
- :issue:`534`: when updating a resource, give a clearer error message if the
  resource ID in the JSON document is not a JSON string.
- :issue:`536`: adds support for single-table inheritance.
- :issue:`540`: correctly support models that don't have a column named "id".
- :issue:`545`: refactors implementation of :class:`.DefaultDeserializer` so
  that it is easier for subclasses to override different subprocedures.
- :issue:`546`: adds support for joined table inheritance.
- :issue:`548`: requests can now use the ``Accept: */*`` header.
- :issue:`559`: fixes bug that stripped attributes with JSON API reserved names
  (like "type") when deserializing resources.
- :issue:`583`: fixes failing tests when using simplejson.
- :issue:`590`: allows user to specify a custom operator for filters.
- :issue:`599`: fixes `unicode` bug using :func:`!urlparse.urljoin` with the
  `future`_ library in resource serialization.
- :issue:`625`: adds schema metadata to root endpoint.
- :issue:`626`: allows the client to request case-insensitive sorting.
- :issue:`630`: correctly respond with timezone-naive :class:`datetime`
  attributes on :http:method:`post` requests when the database doesn't support
  timezones, even if the request included a timezone-aware :class:`datetime`
  attribute.

.. _future: http://python-future.org/

Version 1.0.0b1
---------------

*This is a beta release; these changes will appear in the 1.0.0 release.*

Released on April 2, 2016.

- :issue:`255`: adds support for filtering by PostgreSQL network operators.
- :issue:`257`: ensures additional attributes specified by the user actually exist on
  the model.
- :issue:`363` (partial solution): don't use ``COUNT`` on requests that don't require
  pagination.
- :issue:`404`: **Major overhaul of Flask-Restless to support JSON API**.
- Increases minimum version requirement for ``python-dateutil`` to be strictly
  greater than 2.2 to avoid parsing bug.
- :issue:`331,415`: documents the importance of URL encoding when using the ``like``
  operator to filter results.
- :issue:`376`: add a ``not_like`` operator for filter objects.
- :issue:`431`: adds a ``url_prefix`` keyword argument to the
  :class:`.APIManager` constructor, so one can specify a URL prefix once for
  all created APIs.
- :issue:`449`: roll back the session on any SQLAlchemy error, not just a few.
- :issue:`432,462`: alias relation names when sorting by multiple attributes on a
  relationship.
- :issue:`436,453`: use ``__table__.name`` instead of ``__tablename__`` to infer the
  collection name for the SQLAlchemy model.
- :issue:`440,475`: uses the serialization function provided at the time of
  invoking :meth:`.APIManager.create_api` to serialize each resource correctly,
  depending on its type.
- :issue:`474`: include license files in built wheel for distribution.
- :issue:`501`: allows empty string for `url_prefix` keyword argument to
  :meth:`.APIManager.create_api`.
- :issue:`476`: use the primary key provided at the time of invoking
  :meth:`.APIManager.create_api` to build resource urls in responses.

Older versions

restful/api.py
# -*- coding: utf-8; Mode: Python -*-
#
# Copyright (C) 2011 Lincoln de Sousa <lincoln@comum.org>
# Copyright 2012 Jeffrey Finkelstein <jeffrey.finkelstein@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
    restful.api
    ~~~~~~~~~~~

    Defines a flask model that provides a generic model API to be
    exposed using REST spec.

    This module is intended to be a generic way for creating, updating,
    searching and deleting entries of an sqlalchemy model (declared with
    elixir) through a generic HTTP API.

    All models being exposed by this API are going to be validated using
    formencode validators that should be declared by the user of this
    library.

    :copyright:2011 by Lincoln de Sousa <lincoln@comum.org>
    :copyright:2012 Jeffrey Finkelstein <jeffrey.finkelstein@gmail.com>
    :license: AGPLv3, see COPYING for more details
"""

from collections import namedtuple
import json

from dateutil.parser import parse as parse_datetime
from elixir import session
from flask import abort
from flask import Blueprint
from flask import jsonify
from flask import make_response
from flask import request
from flask.views import MethodView
from sqlalchemy.orm.exc import MultipleResultsFound
from sqlalchemy.orm.exc import NoResultFound
from sqlalchemy.types import Date
from sqlalchemy.types import DateTime

from .search import create_query
from .search import evaluate_functions
from .search import search

__all__ = 'APIManager',


def jsonify_status_code(status_code, *args, **kw):
    """Returns a jsonified response with the specified HTTP status code.

    The positional and keyword arguments are passed directly to the
    :func:`flask.jsonify` function which creates the response.

    """
    response = jsonify(*args, **kw)
    response.status_code = status_code
    return response


class API(MethodView):
    """Provides method-based dispatching for :http:request:`get`,
    :http:request:`post`, :http:request:`patch`, and :http:request:`delete`
    requests, for both collections of models and individual models.

    """

    def __init__(self, model, *args, **kw):
        """Calls the constructor of the superclass and specifies the model for
        which this class provides a ReSTful API.

        ``model`` is the :class:`elixir.entity.Entity` class of the database
        model for which this instance of the class is an API.

        """
        super(API, self).__init__(*args, **kw)
        self.model = model

    # TODO change this to have more sensible arguments
    def _update_relations(self, query, params):
        """Adds or removes models which are related to the model specified in
        the constructor of this class.

        ``query`` is a SQLAlchemy query instance that evaluates to all
        instances of the model specified in the constructor of this class that
        should be updated.

        `params`

            A dictionary with two keys ``add`` and ``remove``. Both of them
            contains a list of items that should be added or removed from
            such a relation.
        """
        fields = []
        cols = self.model.get_columns()
        relations = self.model.get_relations()
        for col in set(relations).intersection(params.keys()):
            submodel = cols[col].property.mapper.class_
            from_request = params[col]

            # Let's add new columns to the relation being managed.
            for subparams in from_request.get('add', ()):
                if 'id' in subparams:
                    subinst = submodel.get_by(id=subparams.pop('id'))
                else:
                    vssubparams = subparams
                    subinst = submodel.get_or_create(**vssubparams)[0]
                for instance in query:
                    getattr(instance, col).append(subinst)

            # Now is the time to handle relations being removed from a field
            # of the instance. We'll do nothing here if there's no id param
            for subparams in from_request.get('remove', ()):
                try:
                    remove = subparams.pop('__delete__')
                except KeyError:
                    remove = False
                if 'id' in subparams:
                    subinst = submodel.get_by(id=subparams['id'])
                else:
                    vssubparams = subparams
                    subinst = submodel.get_by(**vssubparams)
                for instance in query:
                    field = getattr(instance, col)
                    field.remove(subinst)
                if remove:
                    subinst.delete()

            fields.append(col)
        return fields

    def _search(self):
        """Defines a generic search function for the database model.

        If the query string is empty, or if the specified query is invalid for
        some reason (for example, searching for all person instances with), the
        response will be the JSON string ``{"objects": []}``.

        To search for entities meeting some criteria, the client makes a
        request to :http:get:`/api/<modelname>` with a query string containing
        the parameters of the search. The parameters of the search can involve
        filters and/or functions. In a filter, the client specifies the name of
        the field by which to filter, the operation to perform on the field,
        and the value which is the argument to that operation. In a function,
        the client specifies the name of a SQL function which is executed on
        the search results; the result of executing the function is returned to
        the client.

        The parameters of the search must be provided in JSON form as the value
        of the ``q`` request query parameter. For example, in a database of
        people, to search for all people with a name containing a "y", the
        client would make a :http:method:`get` request to ``/api/person`` with
        query parameter as follows::

            q={"filters": [{"name": "name", "op": "like", "val": "%y%"}]}

        If multiple objects meet the criteria of the search, the response has
        :http:status:`200` and content of the form::

            {"objects": [{"name": "Mary"}, {"name": "Byron"}, ...]}

        If the result of the search is a single instance of the model, the JSON
        representation of that instance would be the top-level object in the
        content of the response::

            {"name": "Mary", ...}

        For more information SQLAlchemy functions and operators for use in
        filters, see the `SQLAlchemy SQL expression tutorial
        <http://docs.sqlalchemy.org/en/latest/core/tutorial.html>`_.

        The general structure of request data as a JSON string is as follows::

            {
              "type": "one",
              "order_by": [{"field": "age", "direction": "asc"}],
              "limit": 2,
              "offset": 1,
              "filters":
                [
                  {"name": "name", "val": "%y%", "op": "like"},
                  {"name": "age", "val": [18, 19, 20, 21], "op": "in"},
                  {"name": "age", "op": "gt", "field": "height"},
                  ...
                ],
              "functions":
                [
                  {"name": "sum", "field": "age"},
                  ...
                ]
            }

        For a complete description of all possible search parameters, see
        :ref:`search`.

        """
        # try to get search query from the request query parameters
        try:
            data = json.loads(request.args.get('q', '{}'))
        except (TypeError, ValueError, OverflowError):
            return jsonify_status_code(400, message='Unable to decode data')

        # If the query parameters specify that at least one function should be
        # executed, return the result of executing that function. If no
        # functions are specified, perform a search and return the instances
        # which match that search.
        if 'functions' in data:
            # TODO data['functions'] may be poorly defined here...
            result = evaluate_functions(self.model, data['functions'])
            return jsonify(result)

        # there were no functions specified, so perform a filtered search
        try:
            result = search(self.model, data)
        except NoResultFound:
            return jsonify(message='No result found')
        except MultipleResultsFound:
            return jsonify(message='Multiple results found')

        # create a placeholder for relations of the returned models
        relations = self.model.get_relations()
        deep = dict(zip(relations, [{}] * len(relations)))

        # for security purposes, don't transmit list as top-level JSON
        if isinstance(result, list):
            result = [x.to_dict(deep) for x in result]
            return jsonify(objects=result)
        else:
            return jsonify(result.to_dict(deep))

    # TODO should this exist?
    def _patch_many(self):
        """Updates each of the instances of the model which match the query
        string.

        The ``request.data`` field should be filled with a JSON string that
        contains an object with two fields: query and form. The first one
        should have an object that looks the same as the one passed to the
        ``search`` method. The second field (form) should contain an object
        with all fields that will be passed to the ``update()`` method.

        """
        # TODO this code is common to patch
        try:
            data = json.loads(request.data)
        except (TypeError, ValueError, OverflowError):
            return jsonify_status_code(400, message='Unable to decode data')
        # build the query dictionary from the request query string
        query = create_query(self.model, data)

        # TODO this code is common to patch
        if len(data) == 0:
            return make_response(None, 204)

        # TODO this code is common to patch
        relations = set(self._update_relations(query, data))
        field_list = set(data.keys()) ^ relations
        originalparams = dict((field, data[field]) for field in field_list)

        # Special case: if there are any dates, convert the string form of the
        # date into an instance of the Python ``datetime`` object.
        params = {}
        for key, val in originalparams.iteritems():
            fieldtype = getattr(self.model, key).property.columns[0].type
            if isinstance(fieldtype, Date) or isinstance(fieldtype, DateTime):
                params[key] = parse_datetime(val)
            else:
                params[key] = val

        # Let's update all instances present in the query
        num_modified = 0
        if params:
            num_modified = query.update(params, False)
        session.commit()

        return jsonify(num_modified=num_modified)

    def get(self, instid):
        """Returns a JSON representation of an instance of model with the
        specified name.

        If ``instid`` is ``None``, this method returns the result of a search
        with parameters specified in the query string of the request. If no
        search parameters are specified, this method returns all instances of
        the specified model.

        If ``instid`` is an integer, this method returns the instance of the
        model with that identifying integer. (Implementation note: the
        underlying implementation uses the :func:`elixir.entity.Entity.get_by`
        method.) If no such instance exists, this method responds with
        :http:status:`404`.

        """
        if instid is None:
            return self._search()
        inst = self.model.get_by(id=instid)
        if inst is None:
            abort(404)
        relations = self.model.get_relations()
        deep = dict(zip(relations, [{} for n in range(len(relations))]))
        return jsonify(inst.to_dict(deep))

    def delete(self, instid):
        """Removes the specified instance of the model with the specified name
        from the database.

        Since :http:method:`delete` is an idempotent method according to the
        :rfc:`2616`, this method responds with :http:status:`204` regardless of
        whether an object was deleted.

        """
        inst = self.model.get_by(id=instid)
        if inst is not None:
            inst.delete()
            session.commit()
        return make_response(None, 204)

    def post(self):
        """Creates a new instance of a given model based on request data.

        This function parses the string contained in
        :attr:`flask.request.data`` as a JSON object and then validates it with
        a validator specified in the constructor of this class.

        The :attr:`flask.request.data` attribute will be parsed as a JSON
        object containing the mapping from field name to value to which to
        initialize the created instance of the model.

        After that, it separates all columns that defines relationships with
        other entities, creates a model with the simple columns and then
        creates instances of these submodels and associates them with the
        related fields. This happens only at the first level of nesting.

        Currently, this method can only handle instantiating a model with a
        single level of relationship data.

        """
        # try to read the parameters for the model from the body of the request
        try:
            params = json.loads(request.data)
        except (TypeError, ValueError, OverflowError):
            return jsonify_status_code(400, message='Unable to decode data')

        # Getting the list of relations that will be added later
        cols = self.model.get_columns()
        relations = self.model.get_relations()

        # Looking for what we're going to set on the model right now
        colkeys = cols.keys()
        paramkeys = params.keys()
        props = set(colkeys).intersection(paramkeys).difference(relations)

        # Instantiate the model with the parameters
        instance = self.model(**dict([(i, params[i]) for i in props]))

        # Handling relations, a single level is allowed
        for col in set(relations).intersection(paramkeys):
            submodel = cols[col].property.mapper.class_
            for subparams in params[col]:
                subinst = submodel.get_or_create(**subparams)[0]
                getattr(instance, col).append(subinst)

        # add the created model to the session
        session.add(instance)
        session.commit()

        return jsonify_status_code(201, id=instance.id)

    def patch(self, instid):
        """Updates the instance specified by ``instid`` of the named model, or
        updates multiple instances if ``instid`` is ``None``.

        The :attr:`flask.request.data` attribute will be parsed as a JSON
        object containing the mapping from field name to value to which to
        update the specified instance or instances.

        If ``instid`` is ``None``, the query string will be used to search for
        instances (using the :func:`_search` method), and all matching
        instances will be updated according to the content of the request data.
        See the :func:`_search` documentation on more information about search
        parameters for restricting the set of instances on which updates will
        be made in this case.

        """
        # TODO this code is common to _patch_many
        # if no instance is specified, try to patch many using a search
        if instid is None:
            return self._patch_many()
        
        # try to load the fields/values to update from the body of the request
        try:
            data = json.loads(request.data)
        except (TypeError, ValueError, OverflowError):
            return jsonify_status_code(400, message='Unable to decode data')

        # TODO this code is common to _patch_many
        # If there is no data to update, just return HTTP 204 No Content.
        if len(data) == 0:
            return make_response(None, 204)

        # TODO this code is common to _patch_many
        relations = set(self._update_relations([self.model.get_by(id=instid)],
                                               data))
        field_list = set(data.keys()) ^ relations
        originalparams = dict((field, data[field]) for field in field_list)

        # Special case: if there are any dates, convert the string form of the
        # date into an instance of the Python ``datetime`` object.
        params = {}
        for key, val in originalparams.iteritems():
            fieldtype = getattr(self.model, key).property.columns[0].type
            if isinstance(fieldtype, Date) or isinstance(fieldtype, DateTime):
                params[key] = parse_datetime(val)
            else:
                params[key] = val

        # Let's update field by field
        # TODO can this be made the same as in _patch_many?
        inst = self.model.get_by(id=instid)
        for field in field_list:
            setattr(inst, field, params[field])
        session.commit()

        # return the updated object
        return self.get(instid)


class APIManager(object):
    """TODO fill me in."""

    def __init__(self, app):
        """TODO fill me in.

        ``app`` is the :class:`flask.Flask` object containing the user's Flask
        application.

        """
        self.app = app

    # alternately: def add_api(modelname, readonly=True):
    def create_api(self, model, methods=['GET'], url_prefix='/api'):
        """Creates a ReSTful API interface as a blueprint and registers it on
        the :class:`flask.Flask` application specified in the constructor to
        this class.

        The endpoints for the API for ``model`` will be available at
        ``<url_prefix>/<modelname>``, where ``<url_prefix>`` is the last
        parameter to this function and ``<modelname>`` is the lowercase string
        representation of the model class, as accessed by
        ``model.__name__``. (If any black magic was performed on
        ``model.__name__``, this will be reflected in the endpoint URL.)

        This function must be called at most once for each model for which you
        wish to create a ReSTful API. Its behavior (for now) is undefined if
        called more than once.

        ``model`` is the :class:`elixir.entity.Entity` class for which a
        ReSTful interface will be created. Note this must be a class, not an
        instance of a class.

        ``methods`` specify the HTTP methods which will be made available on
        the ReSTful API for the specified model, subject to the following
        caveats:
        * If :http:method:`get` is in this list, the API will allow getting a
          single instance of the model, getting all instances of the model, and
          searching the model using search parameters.
        * If :http:method:`patch` is in this list, the API will allow updating
          a single instance of the model, updating all instances of the model,
          and updating a subset of all instances of the model specified using
          search parameters.
        * If :http:method:`delete` is in this list, the API will allow deletion
          of a single instance of the model per request.
        * If :http:method:`post` is in this list, the API will allow posting a
          new instance of the model per request.
        The default list of methods provides a read-only interface (that is,
        only :http:method:`get` requests are allowed).

        ``url_prefix`` specifies the URL prefix at which this API will be
        accessible.

        """
        modelname = model.__name__
        methods = frozenset(methods)
        # sets of methods used for different types of endpoints
        no_instance_methods = methods & {'POST'}
        possibly_empty_instance_methods = methods & {'GET', 'PATCH'}
        instance_methods = methods & {'GET', 'PATCH', 'DELETE'}
        # the base URL of the endpoints on which requests will be made
        collection_endpoint = '/{}'.format(modelname)
        instance_endpoint = collection_endpoint + '/<int:instid>'
        # the name of the API, for use in creating the view and the blueprint
        apiname = '{}api'.format(modelname)
        # the view function for the API for this model
        api_view = API.as_view(apiname, model)
        # add the URL rules to the blueprint: the first is for methods on the
        # collection only, the second is for methods which may or may not
        # specify an instance, the third is for methods which must specify an
        # instance
        # TODO what should the second argument here be?
        # TODO should the url_prefix be specified here or in register_blueprint
        blueprint = Blueprint(apiname, __name__, url_prefix=url_prefix)
        blueprint.add_url_rule(collection_endpoint,
                               methods=no_instance_methods, view_func=api_view)
        blueprint.add_url_rule(collection_endpoint, defaults={'instid': None},
                               methods=possibly_empty_instance_methods,
                               view_func=api_view)
        blueprint.add_url_rule(instance_endpoint, methods=instance_methods,
                               view_func=api_view)
        # register the blueprint on the app
        self.app.register_blueprint(blueprint)
